### ISSUE TITLE
Extend default analytics configuration

### ISSUE THREAD

**Issue by swissspidy** [GoogleForCreators/web-stories-wp#2731]
## Feature Description

Right now, the only way to get analytics (e.g. Google Analytics) for stories is via Site Kit.

We currently extend Site Kit's default configuration to add custom triggers for `story-page-visible` and `story-last-page-visible`, as described in [this blog post](https://blog.amp.dev/2019/08/28/analytics-for-your-amp-stories/).

However, there are many more possible triggers, [as the documentation shows](https://github.com/ampproject/amphtml/blob/master/extensions/amp-story/amp-story-analytics.md). For example: `story-focus`, `story-click-through`, `story-audio-muted`, etc.

We should add all those triggers to the default configuration, so users get a great analytics configuration out of the box.

Things that need to be considered for this:

* The `story-page-visible` trigger can be treated as an _event_, or as a _pageview_. The latter could be perceived as [inflated metrics](https://blog.amp.dev/2019/08/28/analytics-for-your-amp-stories/), but perhaps some people would want that 🤷  
  What should be choose? Should the behavior be customizable via a filter or even a UI option?
* Do we need to add `data-vars-*` attributes to elements for the [supported event types](https://github.com/ampproject/amphtml/blob/master/extensions/amp-story/amp-story-analytics.md#variables-as-data-attribute)?  
  These could then be passed to `extraUrlParams` in the config, to pass even more relevant data to the analytics provider.

## Additional Context

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance Criteria

* Add analytics configuration for all [available events/triggers](https://github.com/ampproject/amphtml/blob/master/extensions/amp-story/amp-story-analytics.md) (except the deprecated bookend ones)

## Implementation Brief

* The analytics configuration should be reusable in way that it is not tied to the Site Kit implementation, but if we were to add other providers, we can just include the same configuration.
* New `Analytics` class that accepts a post object as argument.
* New `Analytics::get_configuration()` method that returns the full configuration
* New `Analytics::print()` method that prints the whole `<amp-analytics>` code

**Comment by spacedmonkey** [comment]
I have marked this ticket as blocked #3031 

I have also added an optional follow up ticket in #3130

**Comment by swissspidy** [comment]
The only part that's really blocked by #3031 is printing the analytics code.

Actually extending the default analytics configuration is not blocked by #3031 though.

**Comment by spacedmonkey** [comment]
> The only part that's really blocked by #3031 is printing the analytics code.

But that does mean it blocked right? This ticket can be moved to done until that task is complete. If that is not the case, then maybe the ticket needs to be re-worded, as that is how I read it.

**Comment by swissspidy** [comment]
Removing the following part from the Implementation Brief;

> * In `\Google\Web_Stories\Story_Renderer::maybe_add_analytics`, if `$output` is empty, but a tracking ID is set in the settings, print own analytics code (depends on #3031 / can be covered there if it gets tackled first)

And will create a separate ticket for it.

**Comment by swissspidy** [comment]
I created #3131 for the part that's blocked. Everything in the `Implementation Brief` section can be worked on independently of any existing issues and is not blocked.

**Comment by spacedmonkey** [comment]
So confirm, all we need to do is add these events to the config and the existing amp analytics javascript library will trigger, say on a mute event? Forgive me here, I seem to remember that free versions of Google analytics have a limited number of events that can be fired. I believe events are counted towards hits, which seems to be 10 million a month. Is there a worry with this data, that you burn through those. Should we give user the option, No analytics, basic analytics ( basic page view tracking ) and Advanced analytics ( event based analytics )? At the end of the data, this is their analytics data, they may not which this data tracked. 

To confirm, everything is `Analytics` to keep it tidy right? Will the print method be called on `web_stories_story_head` 

```add_action( 'web_stories_story_head', [ $this, 'print_anayltics' ] );```

Where will this class we created? In discovery? I am guessing you need to pass it the current post. So passing as an object into the constructor doesn't make a tonne of sense to me. 

Should to move this method to the new class as well? 
https://github.com/google/web-stories-wp/blob/2f8e459ad4edb2cee046b99c8a4db186c29db606/includes/Story_Post_Type.php#L489-L545

**Comment by swissspidy** [comment]
> So confirm, all we need to do is add these events to the config and the existing amp analytics javascript library will trigger, say on a mute event?

Correct.

>  Should we give user the option, No analytics, basic analytics ( basic page view tracking ) and Advanced analytics ( event based analytics )?

The idea is that users get the best experience out of the box. Customization might be something for the future, but out of scope for this ticket. Once we add implement this ticket, we can see how well it works.

Note that amp-analytics is not just for Google Analytics. Other providers will use the same configuration too.

> To confirm, everything is `Analytics` to keep it tidy right?

Yep.

> Will the print method be called on `web_stories_story_head`

Not quite. The right hook would be `web_stories_print_analytics`.

**However**, we need to print this code **only** if Site Kit did not already print it.

> Where will this class we created? In discovery?

Maybe? Open to suggestions.

> I am guessing you need to pass it the current post. So passing as an object into the constructor doesn't make a tonne of sense to me.

Can also just pass the current post to the print method, whatever makes most sense.

> Should to move this method to the new class as well?

It shouldn't really be part of `Story_Post_Type`, so yeah why not.

**Comment by swissspidy** [comment]
Resources/ Examples:

https://github.com/ampproject/amphtml/blob/master/extensions/amp-story/amp-story-analytics.md
https://github.com/ampproject/amphtml/blob/b0aff1b9a59d3ca8f10a6ebe4d6eaeda8c826e61/extensions/amp-analytics/0.1/vendors/googleanalytics.json
https://github.com/ampproject/amphtml/blob/0ac07be15dfe6f9b342c7933acb1edaa2d0ba3a6/examples/amp-story/amp-story-branching.html
https://developers.google.com/analytics/devguides/collection/amp-analytics/
